Push linked domains

orangejenny opened the pull request:
##### SUMMARY
https://trello.com/c/x26PQ9nX/26-enterprise-release-management

Allow "pushing" content from master to linked domains, in addition to the current ability for linked domains to "pull" from their master.

Current TODOs:
- [x] Clean up UI on submit (currently runs an ajax request, doesn't clear inputs)
- [x] (maybe) Make a multiselect knockout binding, rather than mixing jQuery and knockout
- [x] Make sure this works properly with the linked project history report

Limitations:
* does not work with remote linked domains
* does not work with multimaster apps

##### FEATURE FLAG
Linked domains

##### RISK ASSESSMENT / QA PLAN
[QA ticket](https://dimagi-dev.atlassian.net/browse/QA-1350)

##### PRODUCT DESCRIPTION
![Screen Shot 2020-05-31 at 11 10 14 PM](https://user-images.githubusercontent.com/1486591/83372921-cbecc300-a394-11ea-9b87-fa181de022ce.png)

orangejenny committed:
Switched UI to use tabs (w=1)

orangejenny committed:
Extracted functions from DomainLinkView.page_context

orangejenny committed:
Added master models, with UI to select models and domains to push to

orangejenny committed:
Added task to do release

orangejenny committed:
Updated UI to not allow form submission until selections are made

orangejenny committed:
Added support for updating models in task

orangejenny committed:
Updated tabs to only show if there's more than one

orangejenny committed:
Added filtering out of remote linked domains

orangejenny committed:
Added support for making and releasing builds for apps

orangejenny committed:
Added comments

orangejenny committed:
Fixed support for pushing reports

Like apps, reports need to be matched up individually with reports in the linked domain.

orangejenny committed:
Fixed lint errors

orangejenny committed:
Fixed lint errors, I'd like to think

orangejenny committed:
Added error when attempting to push an app or report that doesn't exist for linked domain

orangejenny committed:
Added logic to reset form after submission

orangejenny committed:
Added semicolon

orangejenny committed:
Added multiselect knockout binding

orangejenny committed:
Updated js to use knockout instead of jQuery

orangejenny committed:
Fixed trailing comma in arguments list

orangejenny committed:
Fixed lint errors

calellowitz reviewed (approved):
Don't have a ton of context here, but it looks good

orangejenny committed:
Fixed labels

orangejenny committed:
Switched links to buttons

orangejenny committed:
Removed resetForm

The refresh call is breaking the Add All and Remove All buttons.
For the moment, just stop resetting the form when it's submitted.

orangejenny committed:
Removed unused param

orangejenny committed:
Fixed spelling error

orangejenny committed:
Updated email format

orangejenny committed:
Added notify_exception when error is encountered

orangejenny committed:
Fixed tabs for display of projects without links

orangejenny committed:
Updated UI and task to respect case claim flag

orangejenny committed:
Fixed problem with correct tab not showing on page load

orangejenny commented:
@calellowitz This is just about ready, though there are a handful of new commits to deal with qabugs. Could you glance over the last few commits?

orangejenny committed:
Added history update when pushing new content

orangejenny committed:
Added permissions check to DomainLinkView

orangejenny committed:
Updated tab content display logic to match tab display logic

orangejenny committed:
Merge branch 'master' into jls/push-linked-apps

snopoke commented on corehq/apps/linked_domain/tasks.py:
```diff
@@ -1,10 +1,123 @@
+
+
+@task(queue='background_queue')
+def push_models(master_domain, models, linked_domains, build_apps, username):
```
wow, this task is pretty complex - could it be broken up into separate pieces?

orangejenny replied:
yeah, working on it in https://github.com/dimagi/commcare-hq/pull/27878
